Docs - Hamlet Learn (#99)
hamlet-io/docs: docs/learn/layers/product.md

---
sidebar_label: product
title: The Product
---

> A Product is an operational service that has its own lifecycle, often aligning with a organsation Project. 

Reviewer: ```suggestion
A Product is an operational service that has its own life-cycle and represents a single system. 
```

Project gets dicey as that isn't the full lifecycle of a product

We're going to use it to represent a basic phone directory website for our "Acme Inc" organisation.

```bash
hamlet @ ~/cmdb
└─ $ hamlet generate product base --prompt
Enter product id: phonedir
Enter product name [phonedir]:
Enter domain id []: acmeinc
Enter solution id: phonedir
Enter solution name [phonedir]:
Enter environment id: int
Enter environment name [int]: integration
Enter segment id [default]:
Enter segment name [default]:
+-----+------------------+-------------+
|   № | parameter        | value       |
|-----+------------------+-------------|
|   1 | product id       | phonedir    |
|   2 | product name     | phonedir    |
|   3 | domain id        | acmeinc     |
|   4 | solution id      | phonedir    |
|   5 | solution name    | phonedir    |
|   6 | environment id   | int         |
|   7 | environment name | integration |
|   8 | segment id       | default     |
|   9 | segment name     | default     |
+-----+------------------+-------------+
Is everything correct? [y/N]: y
```

> There are a number of default Product types that can be used but here we've chosen the "base" Product - this will give us a well-rounded set of configuration to start us off.

Reviewer: ```suggestion
There are a number of default Product types that we can generate but here we've chosen the "base" Product - this will give us a well-rounded set of configuration to start us off.
```

> Prompt 3 here is asking us what the `Id` value is for the Domain our Product will use. If you recall, we defined our `Domains` object on the Tenant, as it is typically an organisationally owned resource and may be required across a number of Accounts and Products. So reviewing that configuration we see that our new Tenant only has the one domain defined - `acmeinc`.

Reviewer: ```suggestion
Prompt 3 here is asking us what the `Id` value is for the Domain our Product will use. If you recall, we defined our `Domains` object on the Tenant, since domain names need to be unique across the Organisation and may be required across a number of Accounts and Products. So reviewing that configuration we see that our new Tenant only has the one domain defined - `acmeinc`.
```

│   │   │   └── settings
│   │   │       └── shared
│   │   │           └── settings.json
│   │   └── infrastructure
│   │       └── operations
│   │           └── shared
│   │               └── credentials.json
│   └── acmeinc
│       ├── domains.json
│       ├── ipaddressgroups.json
│       └── tenant.json
└── phonedir
    ├── config
    │   ├── product.json
    │   ├── settings
    │   │   ├── integration
    │   │   │   ├── default
    │   │   │   │   └── settings.json
    │   │   │   └── settings.json
    │   │   └── shared
    │   │       └── settings.json
    │   └── solutionsv2
    │       ├── integration
    │       │   ├── default
    │       │   │   └── segment.json
    │       │   └── environment.json
    │       └── shared
    │           └── default
    │               ├── segment.json
    │               └── solution.json
    └── infrastructure
        └── operations
            └── integration
                └── default
                    └── credentials.json
```

## product.json

```bash
hamlet @ ~/cmdb
└─ $ cat ./phonedir/config/product.json
{
    "Product" : {
        "Id" : "phonedir",
        "Name" : "phonedir",
        "Domain" : "acmeinc"
    }
}
```

> Nice and minimal Product configuration here. As is the case for every layer, we define our `Name` and `Id` for it, and our domain association is configured here too.

Reviewer: ```suggestion
This is a basic Product configuration. All layers have an Id and Name field and each layer type adds its own configuration relevant to the product. Here the domain name that will be used by default for any public endpoints we deploy in this product
```

> Our Solution directory looks to have made a created a number of files though, lets take a closer look.

Reviewer: ```suggestion
Our Solution directory looks to have made a created a number of files, lets take a closer look.
```

```bash
.
└── phonedir
    └── config
       └── solutionsv2
           ├── integration
           │   ├── default
           │   │   └── segment.json
           │   └── environment.json
           └── shared
               └── default
                   ├── segment.json
                   └── solution.json
```

The Environment name we chose - `integration` - now has its own directory tree, alongside one called `shared`. This will allow us to separate out configuration intended for a specific environment from that intended to be Environment-layer wide.

> Underneath this, the same has happened for our Segment layer too - however we've opted to go with the `default` segment name. We could create a `shared` Segment directory too, however in this instance we'll only be using the one Segment so its unnecessary.

Reviewer: Might be worth an explanation of an Environment and Segment

Each of the new files defines only it's layer's `Name` and `Id` to distinguish them from eachother, and the solution file also configures an empty `Tiers` object. Though this could have been defined on any Layer, assigning it here means that it will become a part of every Solution for our `phonedir` Product.

```json
{
    "Solution" : {
        "Id" : "phonedir",
        "Name" : "phonedir"
    },
    "Tiers" : {
    }
}

> ```

Reviewer: And an explanation for the Tier as well
